### ISSUE TITLE
Prepare data migration scripts for 2.3

### ISSUE THREAD

**Issue by rimas-kudelis** [avleen/vexim2#59]
Our database structure has been changing over the years, and it looks like we never really had proper migration scripts.

I think we should provide them in 2.3 to make upgrades easier to users of older releases.

**Comment by Udera** [comment]
Question: How do I know which version of vexim is used on my system?
A long time back I install vexim2.2RC1 (sama table layout like 2.2.0?), then I added some patches that were popular at the time: https://www.schirmacher.de/display/INFO/improved+Vexim+frontend+and+bug+fixes
And later I found some more fixes that I added manually.

Would it make sense to keep the old database, then create a new one and copy the data? This won't cause problems if the initial database has been extended and we have a clean database.

**Comment by Udera** [comment]
One more point: you are just deleting the 'clear' field (https://github.com/avleen/vexim2/commit/13e991174e442fb557475ff36eb89622f62bdbba)? And the crypt field used a md5-hash by default? It's not so much better than cleartext, we shouldn't use it anymore.

**Comment by rimas-kudelis** [comment]
2.2RC1, 2.2 and 2.2.1 shipped identical `mysql.sql` files.

Here's how I made these scripts: I downloaded all Vexim releases that I could find, and compared their MySQL databases. The versions I currently have at my disposition are:
- 1.1
- 1.2
- 1.2.1
- 1.3
- 1.5
- 2.0.1
- 2.2RC1
- 2.2
- 2.2.1
It would be really difficult for our migration to support every possible configuration (especially considering the fact that some of us have been using `master` for quite some time), but I think we could support at least some unofficial patches, why not.

On top of that, we could even integrate some of them into our tree, provided that they improve something. :)

Now to go back to your question: I honestly don't know how to figure out which version of Vexim you are using. My assumption was that you would have a tarball somewhere, which would serve as a hint. But in cases like yours, with various patches applied to Vexim, it could very well be that manual migration will be needed.

Regarding your idea to use a second database for migration, I'd say why not. Another option would be to try running the migration from 2.2 and see if that works out. But in any case, you know you should make a backup before starting, don't you? :)

**Comment by rimas-kudelis** [comment]
Regarding the `clear` field: initially what the 2.2 to 2.3 migration did was just remove a couple fields, including this one. I'm currently thinking that we should document migration to 2.3 somewhere, and suggest copying the contents of `clear` column to `crypt` before running this last migration, and then running some other SQL which would encode the passwords. But for this option to be a viable, your `clear` field must be up to date, which is probably not the case if you are using `master`.
I confess I haven't thought it out thoroughly yet. Suggestions are welcome.

**Comment by Udera** [comment]
We could change the password-hash upon the successful login (on vexim). But then we would need to allow different hash-algorithms for authentication at the same time. For md5/sha/.. we could easily use the prefixes `$6$...` to detect the correct scheme, between DES and clear-text will be more difficult. Once the user is authenticated, we create a new password hash.
Courier handles different password schemes at a time as well.

For dovecot-user the situation is even better, Dovecot provides migration scripts: http://wiki2.dovecot.org/HowTo/ConvertPasswordSchemes

**Comment by rimas-kudelis** [comment]
Yeah, I guess we need to update `crypt_password()` to 1) not alter salt in any way, when salt is passed, and 2) add a special case for clear-text auth (and perhaps for SHA auth as well).
I suppose clear-text auth should only work when `$cryptscheme == 'clear'`, in which case encrypted password should be treated as clear-text as well. Assuming that this is a strongly discouraged scenario, I think this would be acceptable.
As for SHA, i find it a bit weird that PHP doesn't have means to generate password hashes in that particular form. I initially thought that it's exim-specific, but it seems not, as at least Dovecot supports it as well. So to work with it, I guess we could check for the {SHA} string at the beginning of the salt, and act accordingly. Although I suppose we should advice against using this algorithm and perhaps even deprecate it in the near future.

**Comment by rimas-kudelis** [comment]
I just looked at the Dovecot wiki page you mentioned. Jeez, I'd say encrypting the value of `clear` field is strongly preferred, unless `clear` is no longer used by a given setup.

**Comment by Udera** [comment]
Before doing a lot of work, maybe we should clarify if there are any users in this situation (no clear-passwords and old salted hashes). Perhaps there is no one or a few who can easily set new passwords for everybody.

Other idea, if migration from DES/SHA1/MD5 (without clear-text) is needed, we implement something in the old version to migrate the passwords. This way, we could start vexim2.3 without these old hashing schemes.

**Comment by rimas-kudelis** [comment]
I don't think we can retro-fit anything into an old version, unless you have a time machine in your basement. :)

I also think it should be simple enough to split vexim_2.2_to_2.3.sql into two separate files, one doing everything but password field, and the other doing the password field. Or to provide an intermediate file which would copy the values of `clear` to `crypt`, which would be optional and only necessary if the administrator wants it (and then add a PHP script which should be run after the migration to encrypt all passwords).

I deally we would have a PHP script responsible for all DB migrations, which could ask the admin some smart questions, or even do some research on its own. The problem with this approach is that it would take significantly more time to implement it (for me at least).

**Comment by rimas-kudelis** [comment]
Oh wait, MySQL also has an `ENCRYPT` function, which calls the `crypt()` of the underlying OS, and which can be salted! This means we can re-encrypt all plain-text passwords within MySQL itself, by using something like the following:
```
UPDATE `users` 
    SET `crypt`=ENCRYPT(`clear`, CONCAT('$6$', MD5(RAND()))) 
    WHERE `crypt`=ENCRYPT(`clear`, `crypt`);
```
before actually dropping the `clear` field.

Two notes about this approach though:
- `clear` and `crypt` might be out of sync (e.g. they are like that in my setup, where I haven't been using `clear` for a while). In this case the original password is beyond recovery, so I guess there's nothing we can do.
- perhaps there could be cases where such automation is undesirable or wouldn't work. Although I guess not?

Also, I wonder if we should still provide a separate migration file for clear-text users? On one hand, this would make our upgrade curve smoother, but on the other hand, is it even worth the time? As far as I'm concerned, we could call that feature unsupported.

**Comment by rimas-kudelis** [comment]
Added a preliminary PR in #72, if anyone would be willing to test.

**Comment by Udera** [comment]
> clear and crypt might be out of sync (e.g. they are like that in my setup, where I haven't been using clear for a while). In this case the original password is beyond recovery, so I guess there's nothing we can do.

First, we should build the migration for default setups and provide information for those who don't use the `clear`-field anymore (uncomment the hashing in the migration script). "Problem" is only, that the new hash scheme is only used, when the users changes his password.

For clear-text users only note in the readme/howto to copy the clear-field in the crypt field and uncomment the password hashing. I wouldn't provide a separate migration script.

**Comment by Udera** [comment]
> UPDATE `users` SET `crypt`=ENCRYPT(`clear`, CONCAT('$6$', MD5(RAND()))) WHERE `crypt`=ENCRYPT(`clear`, `crypt`);

I used this one:
````
UPDATE `users` 
    SET `crypt`=ENCRYPT(`clear`, CONCAT('$6$', MD5(RAND()))) 
    WHERE 1=1;
````
to update all datasets. You can even login with the new calculated hashes. To verify you can output the number of successfull changes:
````
SELECT count(*)
    FROM users
    WHERE `crypt`=ENCRYPT(`clear`, `crypt`);
````
Or you compare that somehow to the total number of lines and then delete the `clear`-field.

**Comment by rimas-kudelis** [comment]
My version only updates `crypt` if it contains the same password as `clear`. I don't think we should be **changing** users' passwords just to promote a better password hashing scheme.

**Comment by Udera** [comment]
Ok, you are right. I only had an old database where passwords weren't in sync any more.

So we can handle the case, where the `clear`-field is still used, all crypt-schemes with `$[:alnum:]$...` should work as well.
Those with `DES` or `SHA1`, with disabled `clear`-field,  will have to keep it, or they set a new default password for everybody. That would be easy to implement. If they need a better migration scheme, we can link to the guide to the dovecot-wiki.

**Comment by Udera** [comment]
Don't forget this fix: https://github.com/vexim/vexim2/pull/147

**Comment by runout-at** [comment]
And: #108

**Comment by runout-at** [comment]
the `users`.`forward` field should be much longer too.
255 -> 4096 ?
how many addresses should go in here?

**Comment by rimas-kudelis** [comment]
Please comment on #72 instead. I suggest we discuss all further changes there, and close this issue as a duplicate. Objections?
